### PR TITLE
#700@minor: Add event listener options.

### PR DESCRIPTION
**Pull request by btea**
rel #700

### CODE REVIEW

File: packages/happy-dom/src/nodes/document/Document.ts [hunk 1]
```diff
@@ -723,7 +723,7 @@ export default class Document extends Node implements IDocument {
 		this._isFirstWriteAfterOpen = true;
 
 		for (const eventType of Object.keys(this._listeners)) {
-			const listeners = this._listeners[eventType];
+			const listeners = this._listeners[eventType].map((listener) => listener.fn);
```
Comment (capricorn86) [inline packages/happy-dom/src/nodes/document/Document.ts:726]: I think we can speed this up a bit by checking directly in the for loop instead as we can avoid the extra loop with the `map()`.

File: packages/happy-dom/src/event/EventTarget.ts [hunk 2]
```diff
@@ -32,7 +41,8 @@ export default abstract class EventTarget implements IEventTarget {
 		listener: ((event: Event) => void) | IEventListener
 	): void {
 		if (this._listeners[type]) {
-			const index = this._listeners[type].indexOf(listener);
+			const _listeners = this._listeners[type].map((l) => l.fn);
```
Comment (capricorn86) [inline packages/happy-dom/src/event/EventTarget.ts:44]: For loops are faster than map(), so I think we are better of with a for loop here, but we could improve performance even further if we store options and listeners in separate objects.

Example:
```typescript
public readonly _options: {
	[k: string]: { once: boolean }[];
} = {};
public readonly _listeners: {
	[k: string]: (((event: Event) => void) | IEventListener)[];
} = {};

public addEventListener(
	type: string,
	listener: ((event: Event) => void) | IEventListener,
	options?: { once: boolean }
): void {
	this._listeners[type] = this._listeners[type] || [];
	this._options[type] = this._options[type] || [];
	this._listeners[type].push(listener);
	this._options[type].push(options || null);
}
```

**Review by capricorn86 — CHANGES_REQUESTED** [review]
Great contribution! :slightly_smiling_face: Just found some minor things.

File: packages/happy-dom/src/nodes/document/Document.ts (hunk 1 above)
Comment (btea) [reply packages/happy-dom/src/nodes/document/Document.ts:726]: Yes, you are right. 👍

File: packages/happy-dom/src/event/EventTarget.ts (hunk 2 above)
Comment (btea) [reply packages/happy-dom/src/event/EventTarget.ts:44]: Great.

**Review by capricorn86 — APPROVED** [review]

**Comment by capricorn86** [comment]
@btea I made a commit with the proposed adjustments for improved performance. Please have a look if you want :slightly_smiling_face:

**Comment by btea** [comment]
It looks very good. 👍

**Pull request merged**